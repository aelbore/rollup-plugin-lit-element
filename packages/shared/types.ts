export type Enforce = 'pre' | 'post'
export type ViteInject = { inject?: boolean }
export type Vite = ViteInject | boolean

export interface Options {
  vite?: Vite
  enforce?: Enforce
  importPackage?: string
  paths?: {
    [from: string]: [string];
  }
}