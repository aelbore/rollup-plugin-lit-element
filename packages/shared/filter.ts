import { createFilter } from '@rollup/pluginutils'

const exclude = [ /node_modules/, /virtual/ ]

export const cssfilter = createFilter('**/*.{css,scss,css?inline,scss?inline}', exclude)
export const tsJsFilter = createFilter('**/*.{js,ts,tsx,jsx,mjs}', exclude)