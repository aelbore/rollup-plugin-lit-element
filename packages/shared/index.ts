export * from './swc'
export * from './utils'
export * from './types'
export * from './filter'