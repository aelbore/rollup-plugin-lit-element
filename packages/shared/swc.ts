import { transformSync, plugins as Plugins, Plugin } from '@swc/core'

export interface SwcOptions {
  plugins?: Plugin[]
  paths?: {
    [from: string]: [string];
  }
}

export const swcTransformer = (
  code: string, 
  id: string, 
  options?: SwcOptions
) => {
  const { paths, plugins } = options || {}
  return transformSync(code, {
    jsc: {
      parser: { 
        syntax: 'typescript', 
        decorators: true, 
        dynamicImport: true 
      },
      target: 'es2022',
      ...(paths ? { paths  }: {})
    },
    filename: id,
    sourceMaps: true,
    isModule: true,
    plugin: Plugins(plugins || [])
  })
}