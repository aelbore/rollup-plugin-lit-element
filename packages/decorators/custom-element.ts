
import { Visitor } from '@swc/core/Visitor.js'
import { getClassDeclaration, hasDecorator } from 'shared'

import * as swc from 'swc-ast-helpers'

import type { Module, Decorator, Program, CallExpression, StringLiteral, ModuleItem, Argument, Identifier } from '@swc/core'

const customElementStatement = (tag: StringLiteral, element: string) => {
  return swc.createExpressionStatement(
    swc.createCallExpression(
      swc.createMemberExpression(swc.createIdentifer('customElements'), swc.createIdentifer('define')),
      [
        { expression: tag },
        { expression: swc.createIdentifer(element) }
      ]
    ))
}

const filterDecorators = (decorators: Decorator[]) => {
  return decorators.filter(decorator => (!(hasDecorator(decorator, 'customElement'))))
}

class CustomElementTransformer extends Visitor { 
  visitModule(m: Module) {
    const moduleItem = getClassDeclaration(m.body)
    const decorator = moduleItem?.decorators?.find(decorator => hasDecorator(decorator, 'customElement')) as Decorator

    if (moduleItem && decorator) {      
      m.body.forEach(content => {
        if (swc.isExportDeclaration(content) && swc.isClasDeclaration(content.declaration)) {
          content.declaration.decorators = filterDecorators(content.declaration.decorators)
        }
        if (swc.isClasDeclaration(content)) {
          content.decorators = filterDecorators(content.decorators)
        }
      })

      const tag = (decorator?.expression as CallExpression)?.arguments[0]?.expression as StringLiteral
      const element = moduleItem.identifier.value
      if (!hasCustomElementStatement(m.body, tag.value, element)) {
        m.body.push(customElementStatement(tag, element))
      }
    }

    return m
  }
}

export function hasCustomElementStatement(items: ModuleItem[], tag: string, element: string) {
  const hasValue = (args: Argument[], value: string) => {
    return args.find(arg => (arg.expression as Identifier).value.includes(value))
  }
  return items.find(item => {
    return swc.isExpressionStatement(item)
      && swc.isCallExpression(item.expression)
      && swc.isMemberExpression(item.expression.callee)
      && swc.isIdentifer(item.expression.callee.object)
      && item.expression.callee.object.value.includes('customElements')
      && swc.isIdentifer(item.expression.callee.property)
      && item.expression.callee.property.value.includes('define')
      && hasValue(item.expression.arguments, tag)
      && hasValue(item.expression.arguments, element)
  })
}

export function customElementTransformer() {
  return (program: Program) => new CustomElementTransformer().visitProgram(program)
}