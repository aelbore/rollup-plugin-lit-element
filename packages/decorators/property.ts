import { Visitor } from '@swc/core/Visitor.js'
import { getClassDeclaration, hasDecorator } from 'shared'

import type { 
  CallExpression, 
  ClassMember, 
  KeyValueProperty, 
  Identifier, Module, 
  ObjectExpression, 
  Program, 
  ClassProperty, 
  Property, 
  SpreadElement,
  Constructor,
  Expression
} from '@swc/core'

import * as swc from 'swc-ast-helpers'

const hasMemberProperty = (member: ClassMember) => 
  swc.isClassProperty(member) && member.decorators.find(decorator => hasDecorator(decorator, 'property'))

const createKeyValueProps = (members: ClassMember[]) => {
  return members.reduce((value, member) => {
    if (swc.isClassProperty(member)) { 
      const decorator = member.decorators.find(decorator => hasDecorator(decorator, 'property'))
      if (decorator) {
        const key = (member.key as Identifier).value
        const args = (decorator?.expression as CallExpression)?.arguments
        const expression = swc.createObjectExpression(
          (args[0]?.expression as ObjectExpression)?.properties 
            ?? [ swc.createKeyValueProperty('type', swc.createIdentifer('String')) ]
        )
        value.push(swc.createKeyValueProperty(key, expression))
      }
    }
    return value
  }, [] as Property[])
}

const createStaticProperties = (members: ClassMember[]) => {
  const props = createKeyValueProps(members)

  const getProperties = (members: ClassMember[]) => {
    return members.find(member => {
      return swc.isClassProperty(member) 
        && swc.isIdentifer(member.key)
        && member.key.value.includes('properties')
    })
  }
  
  const property = (
    getProperties(members) 
      ?? swc.createClassProperty('properties', swc.createObjectExpression([]), { isStatic: true })
  ) as ClassProperty

  const castProp = (prop: Property | SpreadElement) => ((prop as KeyValueProperty).key as Identifier).value

  const propertyValue = property.value as ObjectExpression
  const properties = propertyValue.properties.map(prop => castProp(prop)) || []

  props.forEach(prop => {
    const value = castProp(prop)
    if (!properties.includes(value)) {
      (property.value as ObjectExpression).properties.push(prop)
    }
  })

  return property
}

const updateClassMembers = (members: ClassMember[]) => {
  return members.filter(member => {
     return !(swc.isClassProperty(member) && swc.isIdentifer(member.key) && member.key.value.includes('properties'))
      && !swc.isConstructor(member)
      && !hasMemberProperty(member)
  })
}

const createConstructor = (members: ClassMember[], ctor: Constructor) => {
  const statements = members.map(member => {
    const property = (member as ClassProperty)
    const key = (property.key as Identifier).value

    const left = swc.createMemberExpression(swc.createThisExpression(), swc.createIdentifer(key))
    const right = ((property.value as Identifier) ?? swc.createNullLiteral()) as Expression

    return swc.createExpressionStatement(swc.createAssignmentExpression(left, right))
  })

  const value = ctor ?? 
    swc.createConstructor(
      swc.createBlockStatement([ 
        swc.createExpressionStatement(swc.createCallExpression(swc.createSuper())) 
      ]))

  value.body = swc.createBlockStatement([ ...value.body.stmts, ...statements ])

  return value
}

class PropertyTransformer extends Visitor {
  visitModule(m: Module) {
    const moduleItem = getClassDeclaration(m.body)
    const members = moduleItem?.body?.filter(member => hasMemberProperty(member))

    if (moduleItem && members?.length > 0) {
      const ctor = createConstructor(members, moduleItem.body.find(item => swc.isConstructor(item)) as Constructor)
      const toUpdateMember = (members: ClassMember[]) => [ createStaticProperties(members), ctor, ...updateClassMembers(members) ]
      
      m.body.forEach(content => {
        if (swc.isExportDeclaration(content) && swc.isClasDeclaration(content.declaration)) {
          content.declaration.body = toUpdateMember(content.declaration.body)
        }
        if (swc.isClasDeclaration(content)) {
          content.body = toUpdateMember(content.body)
        }
      })
    }

    return m
  }
}

export function inlinePropertyTransformer() {
  return (program: Program) => new PropertyTransformer().visitProgram(program)
}