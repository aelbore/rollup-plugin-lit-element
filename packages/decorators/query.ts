import { Visitor } from '@swc/core/Visitor.js'
import { getClassDeclaration, hasDecorator, updateMemberDecorators } from 'shared'

import type { 
  CallExpression,
  ClassMember,
  ClassProperty,
  Identifier,
  Module,
  Program 
} from '@swc/core'

import * as swc from 'swc-ast-helpers'

const hasQueryState = (member: ClassMember, key: string) => 
  swc.isClassProperty(member) && member?.decorators?.find(decorator => hasDecorator(decorator, key))

const Query = Object.freeze({
  query: 'querySelector',
  queryAll: 'querySelectorAll'
})

const createGetProperties = (members: ClassMember[]) => {
  return members.map((member: ClassProperty) => {
    const expression = member.decorators[0]?.expression as CallExpression
    const value = (expression?.callee as Identifier).value
    const shadowRoot = swc.createMemberExpression(swc.createThisExpression(), swc.createIdentifer('renderRoot'))
    const statement = swc.createReturnStatement(
      swc.createCallExpression(
        swc.createOptionalChainingExpression(
          swc.createMemberExpression(shadowRoot, swc.createIdentifer(Query[value]))
        ), expression.arguments
      ))
    return swc.createGetter(member.key as Identifier, swc.createBlockStatement([ statement ]))
  })
}

class QueryTransformer extends Visitor {
  visitModule(m: Module) {
    const moduleItem = getClassDeclaration(m.body)
    const members = moduleItem?.body?.filter(member => 
      hasQueryState(member, 'query') || hasQueryState(member, 'queryAll'))

    if (moduleItem && members?.length >  0) {
      const getters = createGetProperties(members)
      const toUpdateMember = (members: ClassMember[]) => {
        const newMembers = updateMemberDecorators(members, Object.keys(Query))
        return [ ...newMembers, ...getters ]
      }
      m.body.forEach(content => {
        if (swc.isExportDeclaration(content) && swc.isClasDeclaration(content.declaration)) {
          content.declaration.body = toUpdateMember(content.declaration.body)
        }
        if (swc.isClasDeclaration(content)) {
          content.body = toUpdateMember(content.body)
        }
      })
    }

    return m
  }
}

export function queryTransformer() {
  return (program: Program) => new QueryTransformer().visitProgram(program)
}
