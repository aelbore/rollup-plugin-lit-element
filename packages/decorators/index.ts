import type { Plugin } from 'rollup'
import type { Enforce } from 'shared'

import { swcTransformer, tsJsFilter } from 'shared'

import { customElementTransformer } from './custom-element'
import { inlinePropertyTransformer } from './property'
import { queryTransformer } from './query'

export default function decorators() {
  return {
    name: 'decorators',
    enforce: 'pre',
    transform(code: string, id: string) {
      if (tsJsFilter(id)) {
        return swcTransformer(code, id, {
          plugins: [ 
            customElementTransformer(), 
            inlinePropertyTransformer(),
            queryTransformer()
          ]
        })
      }
      return null
    }
  } as Plugin & { enforce?: Enforce }
}