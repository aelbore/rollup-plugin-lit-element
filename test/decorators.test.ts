import { parseSync } from '@swc/core'

import { getClassDeclaration, hasDecorator, swcTransformer } from 'shared'
import { customElementTransformer, hasCustomElementStatement } from 'decorators/custom-element'
import { inlinePropertyTransformer } from 'decorators/property'
import { queryTransformer } from 'decorators/query'

describe('decorators', () => {

  it('should transform customElement decorators', () => {
    const content = `
      import { LitElement } from 'lit'
      import { customElement } from 'lit/decorators.js'
      
      @customElement('hello-world')
      export class HelloWorld extends LitElement { }
    `
    
    const { code } = swcTransformer(content, './file.ts', {
      plugins: [ customElementTransformer() ]
    })

    const m = parseSync(code)

    const moduleItem = getClassDeclaration(m.body)
    const decorator = moduleItem?.decorators?.find(decorator => hasDecorator(decorator, 'customElement'))

    expect(decorator).toBeUndefined()
    expect(hasCustomElementStatement(m.body, 'hello-world', 'HelloWorld')).toBeDefined()
    expect(code).toMatchSnapshot()
  })

  it('should transform property decorators', () => {
    const content = `
      import { LitElement } from 'lit'
      import { customElement } from 'lit/decorators.js'
      
      export class HelloWorld extends LitElement { 
        @property({ type: Number }) message: string
      }
    `

    const { code } = swcTransformer(content, './file.ts', {
      plugins: [ inlinePropertyTransformer() ]
    })

    expect(code).toMatchSnapshot()
  })

  it('should transform property decorators existing properties', () => {
    const content = `
      import { LitElement } from 'lit'
      import { customElement } from 'lit/decorators.js'
      
      export class HelloWorld extends LitElement { 
        @property({ reflect: true }) message: string = "World"

        static properties = {
          count: { type: Number }
        }
      }
    `

    const { code } = swcTransformer(content, './file.ts', {
      plugins: [ inlinePropertyTransformer() ]
    })

    expect(code).toMatchSnapshot()
  })

  it('should transform query/queryAll decorators', () => {
    const content = `
      import { LitElement } from 'lit'
      import { query, queryAll } from 'lit/decorators.js'
      
      export class HelloWorld extends LitElement { 
        @query('bto-input') element;
        @queryAll('bto-element') el;
      }
    `

    const { code } = swcTransformer(content, './file.ts', {
      plugins: [ queryTransformer() ]
    })

    expect(code).toMatchSnapshot()
  })

  it('should transform', () => {
    const content = `
      import { LitElement, html } from 'lit'
      import { customElement, property, query } from 'lit/decorators.js'

      @customElement('hello-world')
      export class HelloWorld extends LitElement {
        @property({ type: String }) message: string = 'World'
        @query('i-element') element
        @queryAll('i-input-element') inputElement
      }
    `

    const { code } = swcTransformer(content, './file.ts', {
      plugins: [ 
        customElementTransformer(),
        inlinePropertyTransformer(),
        queryTransformer() 
      ]
    })

    expect(code).toMatchSnapshot()
  })
  
})