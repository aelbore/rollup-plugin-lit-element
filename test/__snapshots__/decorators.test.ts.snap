// Vitest Snapshot v1

exports[`decorators > should transform 1`] = `
"import { LitElement } from 'lit';
export class HelloWorld extends LitElement {
    static properties = {
        message: {
            type: String
        }
    };
    constructor(){
        super();
        this.message = 'World';
    }
    get element() {
        return this.renderRoot?.querySelector('i-element');
    }
    get inputElement() {
        return this.renderRoot?.querySelectorAll('i-input-element');
    }
}
customElements.define('hello-world', HelloWorld);
"
`;

exports[`decorators > should transform customElement decorators 1`] = `
"import { LitElement } from 'lit';
export class HelloWorld extends LitElement {
}
customElements.define('hello-world', HelloWorld);
"
`;

exports[`decorators > should transform property decorators 1`] = `
"import { LitElement } from 'lit';
export class HelloWorld extends LitElement {
    static properties = {
        message: {
            type: Number
        }
    };
    constructor(){
        super();
        this.message = null;
    }
}
"
`;

exports[`decorators > should transform property decorators existing properties 1`] = `
"import { LitElement } from 'lit';
export class HelloWorld extends LitElement {
    static properties = {
        count: {
            type: Number
        },
        message: {
            reflect: true
        }
    };
    constructor(){
        super();
        this.message = \\"World\\";
    }
}
"
`;

exports[`decorators > should transform query/queryAll decorators 1`] = `
"import { LitElement } from 'lit';
export class HelloWorld extends LitElement {
    get element() {
        return this.renderRoot?.querySelector('bto-input');
    }
    get el() {
        return this.renderRoot?.querySelectorAll('bto-element');
    }
}
"
`;
